core/src/storage/collections/btree_map/impls.rs

```diff
@@ -0,0 +1,1872 @@
+/// storage entity. This allows implementations to perform less reads
+/// and writes to the underlying contract storage.
+#[derive(Encode, Decode)]
+#[cfg_attr(feature = "ink-generate-abi", derive(Metadata))]
+pub(super) struct BTreeMapHeader {
+    /// The latest vacant index.
+    next_vacant: Option<u32>,
+    /// The index of the root node.
+    root: Option<u32>,
+    /// The number of elements stored in the map.
+    ///
+    /// # Note
+    ///
+    /// We cannot simply use the underlying length of the vector
+    /// since it would include vacant slots as well.
+    len: u32,
+    /// Number of nodes the tree contains. This is not the number
+    /// of elements!
+    pub(super) node_count: u32,
+}
+
+impl Flush for BTreeMapHeader {
+    #[inline]
+    fn flush(&mut self) {
+        self.next_vacant.flush();
+        self.root.flush();
+        self.len.flush();
+        self.node_count.flush();
+    }
+}
+
+/// A node in the tree.
+///
+/// Each node contains `CAPACITY` keys and values and an edges array over
+/// which children nodes can be linked. Each child has a link to its parent.
+///
+/// Each node is stored as one storage entity. This reduces storage access,
+/// since with each fetch the entire content of a node (all its elements, etc.)
+/// are fetched.
+#[derive(PartialEq, Eq, Encode, Decode)]
+#[cfg_attr(feature = "ink-generate-abi", derive(Metadata))]
+pub struct Node<K, V> {
+    /// This node's parent node.
+    pub parent: Option<u32>,
+
+    /// This node's index into the parent node's `edges` array.
+    /// If, for example, `parent_idx = Some(2)` this refers to the
+    /// second position in the `edges` array of its parent node.
+    pub parent_idx: Option<u32>,
+
+    /// The array storing the keys for a node.
+    pub keys: [Option<K>; CAPACITY],
+
+    /// The array storing the values for a node.
+    pub vals: [Option<V>; CAPACITY],
+
+    /// The pointers to the children of this node.
+    pub edges: [Option<u32>; 2 * B],
+
+    /// Number of elements stored in this node.
+    pub len: u32,
+}
+
+impl<K, V> Flush for Node<K, V>
+where
+    K: Encode + Flush,
+    V: Encode + Flush,
+{
+    #[inline]
+    fn flush(&mut self) {
+        self.parent.flush();
+        self.parent_idx.flush();
+        self.keys.flush();
+        self.vals.flush();
+        self.edges.flush();
+        self.len.flush();
+    }
+}
+
+impl<K, V> Node<K, V> {
+    /// Creates a new `Node`. The node is empty and all fields are
+    /// filled with `None`.
+    pub fn new() -> Self {
+        Node {
+            parent: None,
+            parent_idx: None,
+            keys: Default::default(),
+            vals: Default::default(),
+            edges: [None; 2 * B],
+            len: 0,
+        }
+    }
+
+    /// Returns the number of elements stored in the tree.
+    pub fn len(&self) -> usize {
+        self.len as usize
+    }
+
+    /// Returns the number of edges this node has.
+    pub fn edges(&self) -> usize {
```

Comment:
I'd rename this to `len_edges`. I thought this was returning an iterator over the edges instead tbh.